sympy/physics/mechanics/joint.py

```diff
@@ -0,0 +1,681 @@
+                                  " override apply_joint method in Joint's" +
+                                  " subclass.")
+
+
+class PinJoint(Joint):
+    """
+    Pin (Revolute) Joint.
+
+    It is defined such that the child body rotates with respect to the parent body
+    about the body fixed parent axis through the angle theta. The point of joint
+    (pin joint) is defined by two points in each body which coincides together.
+    They are supplied as parent_point_pos and child_point_pos.
+    In addition, the child's reference frame can be arbitrarily rotated a constant
+    amount with respect to the parent axis by passing an axis in child body which
+    must align with parent axis after the rotation.
+
+    Parameters
+    ----------
+    name: String
+        Name of the joint which makes it unique. Should be different from other
+        joints.
+    parent: Body
+        Instance of Body class serving as the parent in the joint.
+    child: Body
+        Instance of Body class serving as the child in the joint.
+    parent_point_pos: 3 Tuple, Optional
+        Defines the pin joint's point where the parent will be connected to child.
+        3 Tuple defines the values of x, y and z directions w.r.t parent's
+        frame. If it is not supplied, center of mass is added as default.
+    child_point_pos: 3 Tuple, Optional
+        Defines the joint's point where the child will be connected to parent.
+        3 Tuple defines the values of x, y and z directions w.r.t child's frame.
+        If it is not supplied, center of mass is added as default.
+    parent_axis: Vector, Optional
+        Defines the orientation as a vector which must be aligned with child's
+        axis before adding joint. If it is not passed, default is x axis in
+        parent's frame.
+    child_axis: Vector, Optional
+        Defines the orientation as a vector which must be aligned with parent's
+        axis before adding joint. If it is not passed, default is x axis in
+        child's frame.
+
+    Examples
+    --------
+    Adding a Pin Joint which connects center of mass of parent to a point
+    pointed by child.frame.x + child.frame.y w.r.t. child's center of mass.
+    Gravity is along y axis of parent.
+
+    >>> from sympy import Symbol
+    >>> from sympy.physics.mechanics import Body, PinJoint
+    >>> parent = Body('parent')
+    >>> child = Body('child')
+    >>> gravity = Symbol('gravity')
+    >>> l = Symbol('l')
+    >>> child.apply_force(child.mass * gravity * parent.frame.y, child.masscenter)
+    >>> pin_joint = PinJoint('pinjoint', parent, \
+                             child, child_point_pos=(l, l, 0))
+    >>> pin_joint.coordinates
+    [pinjoint_theta(t)]
+    >>> pin_joint.speeds
+    [pinjoint_omega(t)]
+
+    """
+    def __init__(self, name, parent, child, parent_point_pos=None,
+                 child_point_pos=None, parent_axis=None, child_axis=None):
```

Comment:
@moorepants I think this design can be improved , `parent_point_pos` and `child_point_pos` are only need to find the position of center if Joint and if not provided they take pos of masscenter of bodies but take pos(0,0,0) which seems wrong to me.

A better design could be to ask user to pass position of joint rather than position of joint of contact.

`def __init__(self, name, parent, child, joint_pos, parent_axis, child_axis)`

Reply:
I think you have to specify the point on each body relative that body's respective mass center. If you only have 1 joint_pos then you only control the point's location on one body.

Best to start with example problems. For this, try a double pendulum where the joint connecting the two bodies is not at the mass center of either body.